Values, variables, pointers, and references (#2006)
carbon-language/carbon-lang: proposals/p2006.md

### Indexing

Carbon supports indexing that both accesses directly contained storage like an
array and indirect storage like C++'s `std::span`. As a result, the exact
interfaces used for indexing reflect the expression category of the indexed
operand and the specific interface its type implements. This proposal just
updates and refines this design with the new terminology.

> Details: [Indexing](/docs/design/expressions/indexing.md)

### `const`-qualified types

Carbon provides the ability to qualify a type `T` with the keyword `const` to
get a `const`-qualified type: `const T`. This is exclusively an API-subsetting
feature in Carbon -- for more fundamentally "immutable" use cases, value
expressions and bindings should be used instead. Pointers to `const`-qualified
types in Carbon provide a way to reference an object with an API subset that can
help model important requirements like ensuring usage is exclusively by way of a
_thread-safe_ interface subset of an otherwise _thread-compatible_ type.

> Details:
> [`const`-qualified types](/docs/design/values.md#const-qualified-types)

### Interop with `const &` and `const` methods

> Carbon makes value expressions interoperable with `const &` function parameters
> and `const` methods in C++. This is tricky because those constructs in C++ _do_
> allow the address to be taken. As a consequence, Carbon will synthesize an
> object and address for value expressions when necessary for interop. This
> mechanism may also be used as an escape hatch for value expressions, but that is
> [currently not proposed](#value-expression-escape-hatches).
> 
> It may seem risky to expose Carbon's value expressions to `const &` parameters
> in this way, as C++ allows casting away `const`. However, in the absence of
> `mutable` members, casting away `const` does not make it safe to _mutate_
> through a `const &` parameter (or a `const`-qualified method). C++ objects can
> be _declared_ `const`, and those objects are allowed to be implemented in
> readonly memory and assumed to not mutate even if escaped. And `const &`
> parameters can bind to these `const`-declared objects. As a consequence,
> actually mutating through a `const &` parameter (perhaps after casting away
> const) is a fundamentally unsafe operation in C++, and one that
> [breaks in real implementations](https://cpp.compiler-explorer.com/z/KMhTondaK).
> As a consequence, there is strong pressure that C++ code operate with `const &`
> parameters in a way that will be fully compatible with Carbon's value
> expressions.

Reviewer: ```suggestion
Carbon makes value expressions interoperable with `const &` function parameters.
There are two dangers of this approach:

-   Those constructs in C++ _do_ allow the address to be taken, so Carbon will
    synthesize an object and address for value expressions for the duration of
    the C++ call as needed. C++ code already cannot rely on the address of
    `const &` parameters continuing to exist once the call completes.
-   C++ also allows casting away `const`. However, this doesn't allow a method
    to safely _mutate_ a `const &` parameter, except its `mutable` members.

In both cases, correct C++ code should already respect the limitations Carbon
needs to for this interop. The same applies to the implicit `this` parameter of
`const` methods in C++. 
```

        -   [Non-null pointers](#non-null-pointers)
        -   [Syntax-free dereference](#syntax-free-dereference)

<!-- tocstop -->

## Abstract

Introduce a concrete design for how Carbon values, objects, storage, variables,
and pointers will work. This includes fleshing out the design for:

-   The expression categories used in Carbon to represent values and objects,
    how they interact, and terminology that anchors on their expression nature.
-   An expression category model for readonly, abstract values that can
    efficiently support function inputs.
-   A customization system for value expression representations, especially as
    seen on function boundaries in the calling convention.
-   An expression category model for references instead of a type system model.
-   How patterns match different expression categories.
-   How initialization works in conjunction with function returns.
-   Specific pointer syntax, semantics, and library customization mechanisms.
-   A `const` type qualifier for use when the value expression category system
    is too abstracted from the underlying objects in storage.

> ## Problem

Reviewer: Should this link to the appendix?

Author: I'm not sure -- the background is where I currently have that link, which works OK for me. But I'd also be OK moving it if you prefer.

Carbon needs a design for how values, variables, objects, pointers, and
references work within the language. These designs are heavily interdependent
and so they are presented together here. The design also needs to provide a
compelling solution for a wide range of use cases in the language:

> -   Easy to use, and use correctly, function input and in/out parameters.

Reviewer: The items in this list do not have parallel structure. Either they should all look like nouns or they all should have verbs.

Author: I've tried to at least fix some of the parallel structure, but I feel like many of these bullet points are still really awkwardly worded. So far, I've not found obvious better wordings -- ideas here would be welcome.

Reviewer: It looks like you made them into noun phrases, so they probably shouldn't end in periods.

Author: These are noun phrases, but many contain subordinate clauses which contain verbs. The rule excluding a period in Google's style guide, perhaps surprisingly, is just whether the item contains "a verb".

And in the past, I think I've been asked to add periods to the ends of bullet points when, despite not being sentences, they did contain a verb in some subordinate clause.

To be clear, I'm happy to do whatever matches our documentation style guidance. I just thought it actually wanted a period here.

Reviewer: @jonmeow @wolffg could you help us with this documentation style question?

Author: FWIW, this is just in the proposal, so resolving for now. I don't think we need to micro-optimize this.

> -   Clearly separating readonly use cases like function inputs from mutable use

Reviewer: ```suggestion
-   Clearly separating read-only use cases like function inputs from mutable use
```

    cases.

> -   Support for binding local names to both fixed values and mutable variable.

Reviewer: ```suggestion
-   Support for binding local names to both fixed values and mutable variables.
```

> -   Making mutable variable access direct.

Reviewer: Could you clarify what you mean by this?

Author: I have no memory of what past-me meant by this. I'm removing it for now....

-   Indirect (and mutable) access to objects (by way of pointers or references).
-   Extensible models for dereferencing and indexing.

> -   Subsetting method type APIs for readonly input objects.

Reviewer: ```suggestion
-   Subsetting method type APIs for read-only input objects.
```

-   Subsetting method type APIs for indirect access to shared objects in
    thread-compatible ways.
-   Interoperability between function input parameters and C++ `const &`
    parameters.
-   Complex type designs which expose both interior pointers and exterior
    pointers transparently for extended dereferencing or indexing.

### Conceptual integrity between local variables and parameters

An additional challenge that this design attempts to address is retaining the
conceptual integrity between local variables and parameters. Two of the most
fundamental refactorings in software engineering are _inlining_ and _outlining_
of regions of code. These operations introduce or collapse one of the most basic

> abstraction boundaries in the language: functions. When performing this
> refactoring, there is a need to translate between local variables and parameters
> in both directions. In order to ensure these translations are unsurprising and

Reviewer: Passive voice
```suggestion
abstraction boundaries in the language: functions. These refactorings translate
between local variables and parameters
in both directions. In order to ensure these translations are unsurprising and
```

don't face significant expressive gaps or behavioral differences, it is
important to have strong semantic consistency between local variables and
function parameters. While there are some places that these need to differ,
there should be a strong overlap of the core facilities, design, and behavior.

> ## Background

Reviewer: Possibly add "The C++ rvalue lifetime disaster" https://www.youtube.com/watch?v=Tz5drzXREW0 ?

Author: Done.

> 

Reviewer: Please include the issues being implemented by this proposal such as [#523](https://github.com/carbon-language/carbon-lang/issues/523), plus relevant prior proposals, if any.

Author: Added this issue and a prior proposal. I'm not recalling others, but I'll try to go back through again. Let me know if you have any in mind.

operand and the specific interface its type implements. This proposal just
updates and refines this design with the new terminology.

> Details: [Indexing](/docs/design/expressions/indexing.md)

### `const`-qualified types

Carbon provides the ability to qualify a type `T` with the keyword `const` to
get a `const`-qualified type: `const T`. This is exclusively an API-subsetting
feature in Carbon -- for more fundamentally "immutable" use cases, value
expressions and bindings should be used instead. Pointers to `const`-qualified
types in Carbon provide a way to reference an object with an API subset that can
help model important requirements like ensuring usage is exclusively by way of a
_thread-safe_ interface subset of an otherwise _thread-compatible_ type.

> Details:
> [`const`-qualified types](/docs/design/values.md#const-qualified-types)

### Interop with `const &`

> TODO

### Customization

> TODO

## Rationale

-   Pointers are a fundamental components of all modern computer hardware --
    they are abstractly random-access machines -- and being able to directly
    model and manipulate this is necessary for

>     [performance-critical software](/docs/project/goals.md#performance-critical-software).

Reviewer: Question: Could you clarify how the situation with pointers is different from the situation with "goto", which are also a fundamental building block of assembly instructions?

I feel like there is a crowd of people who are very concerned that pointers are the source of bugs and unsafety, and would prefer a language without them.

Author: I'm happy to add this, but before I do -- is there a specific place that this would fit for you in the overall document? That would help me out, as right now I think I know what I want to say to compare and contrast them, but I'm not sure how to make it fit or integrate it into the whole.

Reviewer: Doing it in this section would be fine, or you could explain it in a "no first-class pointers/reference" alternative which would compare with the Val approach using `inout` parameters and so on.

Author: Added it here so I could focus more on the goto side.

Let me know if you'd like an alternative as well.

-   Simplifying the type system by avoiding both pointers and references is
    expected to make
    [code easier to read, understand, and write](/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write).
-   Creating space in both the syntax and type system to introduce ownership and
    lifetime information is important to be able to address long term
    [safety](/docs/project/goals.md#practical-safety-and-testing-mechanisms)
    needs.
-   Pointers are expected to be deeply familiar to C++ programmers and easily
    [interoperate with C++ code](/docs/project/goals.md#interoperability-with-and-migration-from-existing-c-code).

## Alternatives considered

> TODO - this remains very incomplete

### Immutable value escape hatch

> We could provide immediately the escape hatch of a syntax to unsafely take the

Reviewer: ```suggestion
We could provide the escape hatch of a syntax to unsafely take the
```

treated as such in several contexts.

Second, this added complexity would reside exactly in the position of the type
system where additional safety complexity may be needed. We would like to leave
this area (pointers and references to non-local objects) as simple and minimal
as possible to ease the introduction of important safety features going forward
in Carbon.

### Automatic dereferencing

One way to make pointers behave very nearly the same as references without
adding complexity to the type system is to automatically dereference them in the
relevant contexts. This can, if done carefully, preserve the ability to
distinguish between the pointer and the pointed-to object while still enabling
pointers to be seamlessly used without syntactic overhead as L-values.

However, this makes code dereferencing a pointer and performing a non-local and
potentially mutating operation visually indistinct. Having visual markers for
this arguably provides some readability improvement for some people, but is
noise and a distraction for others. Reasonable judgement calls about which
direction to prefer may differ, but Carbon's principle of preferring lower

> context sensitivity leans (slightly) toward explicit dereferencing instead.

Reviewer: Link to the principle: https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/principles/low_context_sensitivity.md

It is worth noting that there are existing languages that use exactly this or an
extremely similar pattern such as Rust. It is also relatively easy to imagine
moving from this proposal toward automatic dereferencing in the future as it
builds on the same core type system primitives.

### Exclusively using references

While framed differently, this is essentially equivalent to automatic
dereferencing of pointers. The key is that it does not add both options to the
type system but addresses the syntactic differences separately and uses
different operations to distinguish between the reference and the referenced

> object when necessary.

Reviewer: Explain why this alternative isn't chosen?

Author: This was supposed to be part of the equivalence to the previous. I've made it more explicit, does that work?

### Alternative syntaxes for locals

> TODO: needs to be rewritten

**Syntactic alternatives not currently proposed:**

-   It is appealing to consider the `val` introducer instead of `let` given that
    these are expected to be called "values". However, there is some concern
    over the visual similarity of `val` and `var`, and if pronounced as written
    instead of as "value" ar "variable", the difference relies on a subtle

>     consonant distinction not present in some widely used spoken language. That

Reviewer: ```suggestion
    consonant distinction not present in some widely used spoken languages. That
```

or

```suggestion
    consonant distinction not present in a widely used spoken language. That
```

that should be covered by the proposed design.

### `const` references versus `const` itself

C++ provides overlapping but importantly separable semantic models which
interact with `const` references.

1. An _immutable view_ of a value
2. A _thread-safe interface_ of a
   [thread-compatible type](https://abseil.io/blog/20180531-regular-types#:~:text=restrictions%20or%20both.-,Thread-compatible,-%3A%20No%20concurrent%20call)

Some examples of the immutable view use case are provided below. These include
`const` reference parameters and locals, as well as `const` declared local and
static objects.

```
void SomeFunction(const int &id) {
  // Here `id` is an immutable view of some value provided by the caller.
}

void OtherFunction(...) {
  // ...

  const int &other_id = <some-runtime-expression>;

  // Cannot mutate `other_id` here either, it is just a view of the result of
  // `<some-runtime-expression>` above. But we can pass it along to another
  // function accepting an immutable view:
  OtherFunction(other_id);

  // We can also pass ephemeral values:
  OtherFunction(other_id + 2);

  // Or values that may be backed by read-only memory:
  static const int fixed_id = 42;

>   OtherFunction(fixed_id);

Reviewer: ```suggestion
  SomeFunction(other_id);

  // We can also pass ephemeral values:
  SomeFunction(other_id + 2);

  // Or values that may be backed by read-only memory:
  static const int fixed_id = 42;
  SomeFunction(fixed_id);
```

significant syntactic overhead.

#### Member and subobject accessors

Another common use of references is in returns from member functions to provide
access to a member or subobject, whether const or mutable. This particular use
case is worth calling out specially as it has an interesting property: this is
often not a fully indirect access. Instead, it is often simply selecting a
particular member, field, or other subobject of the data structure. As a
consequence, making subsequent access transparent seems especially desirable.

However, it is worth noting that this particular use case is also an especially
common source of lifetime bugs. A classic and pervasive example can be seen when
calling such a method on a temporary object. The returned reference is almost
immediately invalid.

#### Non-null pointers

A common reason for using mutable references outside of what has already been
described is to represent _non-null pointers_ with enforcement in the type
system. Because the canonical pointer types in C++ are allowed to be null,
systems that forbid a null in the type system use references to induce any null

> checks to be as early as possible. This effects a

Reviewer: ```suggestion
checks to be as early as possible. This affects a
```
though 
```suggestion
checks to be as early as possible. This causes a
```
is probably better.

null pointers, both moving the error closer to its cause logically and
increasing the chance of moving earlier in the development process by making it
a static property enforced at compile time.

References are imperfectly suited to modeling non-null pointers because they are
missing many of the fundamental properties of pointers such as being able to
rebind them, being able to take their address, etc. Also, references cannot be
safely made `const` in the same places that pointers can because that might
unintentionally change their semantics by allowing temporaries or extending
lifetimes.

#### Syntax-free dereference

Beyond serving as a non-null pointer, the other broad use case for references is
to remove the syntactic overhead of taking an address and dereferencing
pointers. In other words, they provide a way to have _syntax-free dereferences_.
Outside of function parameters, removing this distinction may provide a
genericity benefit, as it allows using the same syntax as would be used with
non-references. In theory code could simply use pointers everywhere, but this
would add syntactic overhead compared to local variables and references. For

> immutable code, the syntactic overhead seems unnecessary and unhelpful. However,

Reviewer: > For immutable code,

Is the code what is immutable here, or is it code that is not performing mutation?

Author: Switched to "immutable accesses".

    -   [`const` references versus `const` itself](#const-references-versus-const-itself)
    -   [Pointers](#pointers)
    -   [References](#references)
        -   [Special but critical case of `const T&`](#special-but-critical-case-of-const-t)
        -   [R-value references and forwarding references](#r-value-references-and-forwarding-references)
        -   [Mutable operands to user-defined operators](#mutable-operands-to-user-defined-operators)
        -   [User-defined dereference and indexed access syntax](#user-defined-dereference-and-indexed-access-syntax)
        -   [Member and subobject accessors](#member-and-subobject-accessors)
        -   [Non-null pointers](#non-null-pointers)
        -   [Syntax-free dereference](#syntax-free-dereference)

<!-- tocstop -->

## Abstract

Introduce a concrete design for how Carbon values, objects, storage, variables,
and pointers will work. This includes fleshing out the design for:

-   The expression categories used in Carbon to represent values and objects,
    how they interact, and terminology that anchors on their expression nature.
-   An expression category model for readonly, abstract values that can

>     efficiently support function inputs.

Reviewer 2: ```suggestion
    efficiently support input to functions.
```
(I read this as meaning "functions as inputs to abstract values" at first.)

covered in the updated content in the design, and each section links into the
relevant content there. While this overview both duplicates and summarizes
content, it isn't intending anything different from the updates to the design
content, and the design content should be considered authoritative as it will
also continue to be maintained going forward.

### Values, objects, and expressions

Carbon has both abstract _values_ and concrete _objects_. Carbon _values_ are
things like `42`, `true`, and `i32` (a type value). Carbon _objects_ have
_storage_ where values can be read and written. Storage also allows taking the
address of an object in memory in Carbon.

Both objects and values can be nested within each other. For example
`(true, true)` is both a value and also contains two sub-values. When a
two-tuple is stored somewhere, it is both a tuple-typed object and contains two
subobjects.

> Details:
> [Values, objects, and expressions](/docs/design/values.md#values-objects-and-expressions)

Expressions are categorized in a way that explains how they produce values or
refer to objects:

-   [_Value expressions_](/docs/design/values.md#value-expressions) produce

>     abstract, read-only _values_ that cannot be modified or have its address

Reviewer 2: ```suggestion
    abstract, read-only _values_ that cannot be modified or have their address
```
(plurality mismatch)

    are used to model function returns, which can construct the returned value
    directly in the caller's storage.

> Details: [Expression categories](/docs/design/values.md#expression-categories)

### Patterns, `var`, `let`, and local variables

Patterns are by default _value patterns_ and match _value expressions_, but can
be introduced with the `var` keyword to create a _variable pattern_ that has
_storage_ and matches _initializing expressions_. Names bound in value patterns
become value expressions, and names bound in a variable pattern become _durable
reference expressions_ referring to an object in that pattern's storage.

Local patterns can be introduced with `let` to get the default behavior of a
readonly pattern, or they can be directly introduced with `var` to form a
variable pattern and declare mutable local variables.

> Details:
> [Binding patterns and local variables with `let` and `var`](/docs/design/values.md#binding-patterns-and-local-variables-with-let-and-var)

### Pointers, dereferencing, and references

Pointers in Carbon are the primary mechanism for _indirect access_ to storage
containing some object. Dereferencing a pointer forms a

> [_reference expression_](/docs/design/values.md#reference-expressions) to the

Reviewer 2: Is this always a durable reference expression?

Author: Yes, and updated.

        example, nesting a `var` pattern within `const` might be especially
        surprising.

Ultimately, the overwhelming most popular introducer for immutable local values
across popular languages that have such a distinction is `let`. Using that makes
Carbon unsurprising in the world of programming languages which is a good thing.
Using `var` to help signify the allocation of storage and given it also having
widespread usage across popular languages.

## Appendix: background, context, and use cases from C++

This appendix provides an examination of C++'s fundamental facilities in the
space involving `const` qualification, references (including R-value
references), and pointers. Beyond the expression categorization needed to
provide a complete model for the language, these use cases help inform the space
that should be covered by the proposed design.

### `const` references versus `const` itself

C++ provides overlapping but importantly separable semantic models which
interact with `const` references.

1. An _immutable view_ of a value
2. A _thread-safe interface_ of a [thread-compatible type][]

[thread-compatible type]:

>     https://abseil.io/blog/20180531-regular-types#:~:text=restrictions%20or%20both.-,Thread-compatible,-%3A%20No%20concurrent%20call

Reviewer 2: ```suggestion
    https://abseil.io/blog/20180531-regular-types#:~:text=restrictions%20or%20both,No%20concurrent%20call
```
That link doesn't work in my browser; this is what Chrome produces for me, which does seem to work.

}

void OtherFunction(...) {
  // ...

  const int &other_id = <some-runtime-expression>;

  // Cannot mutate `other_id` here either, it is just a view of the result of
  // `<some-runtime-expression>` above. But we can pass it along to another
  // function accepting an immutable view:
  SomeFunction(other_id);

  // We can also pass ephemeral values:
  SomeFunction(other_id + 2);

  // Or values that may be backed by read-only memory:
  static const int fixed_id = 42; SomeFunction(fixed_id);
}
```

The _immutable view_ `id` in `SomeFunction` can be thought of as requiring that
the semantics of the program be exactly the same whether it is implemented in
terms of a view of the initializing expression or a copy of that value, perhaps
in a register.

The implications of the semantic equivalence help illustrate the requirements:

-   The input value must not change while the view is visible, or else a copy
    would hide those changes.
-   The view must not be used to mutate the value, or those mutations would be
    lost if made to a copy.
-   The identity of the object must not be relevant, or else inspection of its
    address would reveal whether a copy was used.

> Put differently, these restrictions makes a copy valid under the

Reviewer 2: ```suggestion
Put differently, these restrictions make a copy valid under the
```

> Details:
> [`const`-qualified types](/docs/design/values.md#const-qualified-types)

### Interop with `const &` and `const` methods

Carbon makes value expressions interoperable with `const &` function parameters
and `const` methods in C++. This is tricky because those constructs in C++ _do_
allow the address to be taken. As a consequence, Carbon will synthesize an
object and address for value expressions when necessary for interop. This
mechanism may also be used as an escape hatch for value expressions, but that is
[currently not proposed](#value-expression-escape-hatches).

This interop is based on the fact that these constructs in C++ can bind to
`const`-declared objects which strongly pressure code to be compatible even with

> Carbon's more restrictive value expression semantics.

Reviewer: I feel like I don't fully understand this sentence. Could you explain the steps in the reasoning a bit more explicitly for me?

Author: Sure, may have provided too much detail now.

### Interop with `const &` and `const` methods

Carbon makes value expressions interoperable with `const &` function parameters.
There are two dangers of this approach:

-   Those constructs in C++ _do_ allow the address to be taken, so Carbon will
    synthesize an object and address for value expressions for the duration of
    the C++ call as needed. C++ code already cannot rely on the address of
    `const &` parameters continuing to exist once the call completes.
-   C++ also allows casting away `const`. However, this doesn't allow a method
    to safely _mutate_ a `const &` parameter, except its `mutable` members.

In both cases, correct C++ code should already respect the limitations Carbon

> needs to for this interop. The same applies to the implicit `this` parameter of

Reviewer: Sorry, my suggestion had a mistake

```suggestion
needs for this interop. The same applies to the implicit `this` parameter of
```

mechanism may also be used as an escape hatch for value expressions, but that is
[currently not proposed](#value-expression-escape-hatches).

It may seem risky to expose Carbon's value expressions to `const &` parameters
in this way, as C++ allows casting away `const`. However, in the absence of
`mutable` members, casting away `const` does not make it safe to _mutate_
through a `const &` parameter (or a `const`-qualified method). C++ objects can
be _declared_ `const`, and those objects are allowed to be implemented in
readonly memory and assumed to not mutate even if escaped. And `const &`
parameters can bind to these `const`-declared objects. As a consequence,
actually mutating through a `const &` parameter (perhaps after casting away
const) is a fundamentally unsafe operation in C++, and one that
[breaks in real implementations](https://cpp.compiler-explorer.com/z/KMhTondaK).
As a consequence, there is strong pressure that C++ code operate with `const &`
parameters in a way that will be fully compatible with Carbon's value

> expressions.

Reviewer: The case I'm familiar with of mutating a value after casting away `const` is in a situation where you start with a non-`const` object, call a `const` method that returned a `const` reference to some sub-object, cast that `const` away, and then mutate. Basically: casting away `const` that was added by the code in question in order to treat a `const` method as polymorphic on `const`. I'm guessing this is a pattern that we won't support, but we don't expect it to be an issue at the C++ interop boundary?

Author: I think it's not relevant for *this* aspect of interop which is focused on passing *in* a value expression to a `const &` parameter or the object of a `const` method, and whether that will be safe.

(I know we also discussed this a bit over lunch, not sure if that helped clarify anything.)

Reviewer: The structure here is a bit weird. The "details link" below only covers the rationale for why the previous paragraph is safe. It would be better for *that* to also explain why casting away `const` isn't a new danger, then make this section short and link to it. I'll try taking a stab at it.

Author: Accepted all the edits, so I think this is good now.

these differences:

-   C pointers can be null, the ["billion dollar mistake"][null-mistake],
    Carbon's cannot be. Carbon requires null-ness to be explicitly modeled in
    the type system and dealt with prior to dereference.
-   C pointers can be indexed even when not pointing to an array. Carbon's
    pointers are to singular objects. Carbon handles indexing with a separate
    construct and types to distinguish arrays where indexing is semantically
    meaningful from the cases where it is not.
-   C pointers support deeply surprising indexing patterns such as `3[pointer]`.
    Carbon restricts to unsurprising indexing.
-   C pointers don't enforce lifetime or any other memory safety properties.
    While Carbon does not yet support these deeper protections, tackling memory
    safety is an explicit plan of Carbon's and may yet result in changes to the
    pointer type to reflect the safety risks posed by them.

[null-mistake]:
    https://www.infoq.com/presentations/Null-References-The-Billion-Dollar-Mistake-Tony-Hoare/

While these are described in terms of C's pointers, C++ inherits these without
meaningful improvement from C. The core point here is that while Carbon is

> building in pointers to its foundations, it is building in the least problematic

Reviewer 3: ```suggestion
building pointers into its foundations, it is building in the least problematic
```

into its design -- pointers in C and C++ have been the root of systematic
security and correctness bugs. There is an amazing wealth of "gotchas" tied to
pointers that it can seem unreasonable to build Carbon on top of these
foundations. By analogy, the `goto` control-flow construct is similarly
considered _deeply_ problematic and there is a wealth of literature in computer
science and programming language design on how and why to build languages on top
of structured control flow instead. It would be very concerning to build `goto`
into the fundamental design of Carbon as an integral and pervasive component of
its control flow. However, there are two important distinctions between pointers
and `goto`.

First, C and C++ pointers are _very_ different from Carbon pointers, and their
rightly earned reputation as a source of bugs and confusion stems precisely from
these differences:

-   C pointers can be null, the ["billion dollar mistake"][null-mistake],
    Carbon's cannot be. Carbon requires null-ness to be explicitly modeled in
    the type system and dealt with prior to dereference.
-   C pointers can be indexed even when not pointing to an array. Carbon's

>     pointers are to singular objects. Carbon handle indexing with a separate

Reviewer: ```suggestion
    pointers are to singular objects. Carbon handles indexing with a separate
```

-   Visually, `*` is used for both pointers and multiplication.
-   Interacts poorly with prefix array type syntaxes.
-   Looks like a C++ pointer but likely has different semantics.

**Prefix `*`:**

```
var p: *i32 = &i;
```

Advantages:

-   Used by Rust, Go, Zig,
    [Jai](https://github.com/BSVino/JaiPrimer/blob/master/JaiPrimer.md#memory-management).
-   Go in particular has good success transitioning C/C++ programmers to this
    syntax.
-   Allows reading types left-to-right.

Disadvantages:

-   "Pointer to T" and "Dereference T" would not be distinguished grammatically,
    could particularly be a problem in template code
-   `*` used for both pointers and multiplication
-   Uncanny valley -- it is very close to C++'s syntax but not quite the same.

**Prefix `&`:**

```
var p: &i32 = &i;
```

Advantages:

-   "Type of a pointer is a pointer to the type"
-   Allows reading types left-to-right.

Disadvantages:

-   Visual ambiguity:

    ```
    let X:! type = i32;
    // Can't actually write this, but there is a visual collision between
    // whether this is the address of `X` or pointer-to-`i32`.
    var y: auto = &X;
    ```

> 

Reviewer: Looks like something about this is confusing prettier and github's syntax highlighting.

Author: Not sure I can change anything to fix it. I think it renders correctly? I've tried one tweak, but unsure if it will work.

**Prefix `^`:**

```
var p: ^i32 = &i;
```

Advantages:

-   Used by the Pascal family of languages like
    [Delphi](http://rosettacode.org/wiki/Pointers_and_references#Delphi), and a
    few others like
    [BBC_BASIC](http://rosettacode.org/wiki/Pointers_and_references#BBC_BASIC)
-   `^` looks pointy
-   `^` operator is not heavily used otherwise (as a binary op it could be
    bit-xor or raise-to-power).
-   Allows reading types left-to-right.

**`Ptr(T)`:**

```
var p: Ptr(i32) = &i;
```

Advantages:

-   It is simple and unambiguous.
-   Better supports a variety of pointer types.
-   Allows reading types left-to-right.

Disadvantages:

-   Ends up making common code verbose.
-   Not widely used in other languages.

#### Pointer dereference syntax alternatives

**Prefix `*`:**

```

> *p = *p \* *p + (*q)[3] + r->x;

Reviewer: ```suggestion
*p = *p * *p + (*q)[3] + r->x;
```

```

Advantages:

-   Matches C/C++, and will be familiar to our initial users.
-   When _not_ composed with other operations, prefix may be easier to recognize
    visually than postfix.

Disadvantages:

-   `*` is used for both pointers and multiplication.
-   Need parens or operator `->` to resolve precedence issues when composing
    with postfix and infix operations, which is common.

**Postfix `^`:**

```

> p^ = p^ \* p^ + q^[3] + r^.x;

Reviewer: ```suggestion
p^ = p^ * p^ + q^[3] + r^.x;
```

```

Advantages:

-   Generally fewer precedence issues.
-   No need for `->` operator.
-   Used by Pascal family of languages.

Disadvantages:

-   Not familiar to C/C++ users.
-   Potential conflict with `^` as xor or exponentiation.
-   Non-ergonomic: would be very frequently typed, but also a stretching motion
    for typing.

**Postfix `[]`:**

```

> p[] = p[] \* p[] + q[][3] + r[].x;

Reviewer: ```suggestion
p[] = p[] * p[] + q[][3] + r[].x;
```

**Postfix `^`:**

```
p^ = p^ \* p^ + q^[3] + r^.x;
```

Advantages:

-   Generally fewer precedence issues.
-   No need for `->` operator.
-   Used by Pascal family of languages.

Disadvantages:

-   Not familiar to C/C++ users.
-   Potential conflict with `^` as xor or exponentiation.
-   Non-ergonomic: would be very frequently typed, but also a stretching motion
    for typing.

**Postfix `[]`:**

```
p[] = p[] \* p[] + q[][3] + r[].x;
```

Advantages:

-   Generally fewer precedence issues.
-   No need for `->` operator.
-   Used by Swift.

Disadvantages:

-   Two characters is a bit heavier visually.
-   Possibly confusing given that plain pointers are used for non-array
    pointees.

Maybe should pair this with prefix `[]` to make pointer types? Would also need
to distinguish slice types and maybe dynamically-sized-array types.

#### Pointer syntax conclusion

Ultimately, the decision in [#523] was to keep Carbon's syntax familiar to
C++'s. The challenges that presents and advantages of changes weren't sufficient
to overcome the advantage of familiarity and for the specific challenges we have
effective solutions.

### Alternative syntaxes for locals

> > TODO: needs to be rewritten

Reviewer: Note that this is still here.

Author: Thanks for the reminder, done. I hope the result makes sense, happy to augment or change it.

static objects.

```

void SomeFunction(const int &id) { // Here `id` is an immutable view of some
value provided by the caller. }

void OtherFunction(...) { // ...

const int &other_id = <some-runtime-expression>;

// Cannot mutate `other_id` here either, it is just a view of the result of //
`<some-runtime-expression>` above. But we can pass it along to another //
function accepting an immutable view: SomeFunction(other_id);

// We can also pass ephemeral values: SomeFunction(other_id + 2);

// Or values that may be backed by read-only memory: static const int fixed_id =

> 42; SomeFunction(fixed_id); }

Reviewer: Looks like prettier autoformat got confused and garbled the formatting here and below.

Reviewer: This is still garbled. Please undo the changes to this section made in https://github.com/carbon-language/carbon-lang/pull/2006/commits/6c1aed031b2710231ef6e22edc3804bb29dbe57e .

Reviewer: In that commit, I'm referring to the changes to file `p2006.md` starting with `void SomeFunction(`, prior to that commit they were lines 451-460 and after that commit lines 677-682.

In the most recent commit, I see garbling in lines 721-737, 761-776, 784-791 in `p2006.md`.

Author: I think this is all fixed now.

### Alternative syntaxes for locals

Several other options for declaring locals were considered, but none ended up
outweighing the proposed option on balance.

-   `let` and `let mut`, based on the Rust names
    -   While nicely non-inventive, there was both concern that `mut` didn't as
        effectively communicate the requirement of storage and concern around
        not being as obviously or unambiguously a good default with `let`.
    -   The `mut` keyword feels more fitting in its use with mutable borrows
        than as a declaration introducer the way it would work in Carbon. For

>         example, it doesn't convey a desired aspect here of allocating storage.

Reviewer: This seems to be saying "doesn't convey allocating storage" twice, and not saying we wanted the two cases to be equally easy to say.

Author: Tried to reword a bit and add the other aspect.

> -   `val` and `var`

Reviewer: Missing the pronunciation concern.

Author: Hopefully added, let me know if I'm mischaracterizing it though.

Using `var` to help signify the allocation of storage and given it also having
widespread usage across popular languages.

### Make `const` a postfix rather than prefix operator

Using a prefix operator for `const` and a postfix operator for `*` causes them
to require parentheses for complex nesting. We could avoid that by using a
postfix `const` and this would also allow more combinations in Carbon to be
valid in C++ with the same meaning as Carbon such as `T* const`.

This direction isn't pursued because:

-   We expect pointers-to-`const` to be significantly more common in code than
    `const`-pointers-to-non-`const`, even more-so than is already the case in
    C++. And for `pointers-to-const`, we lean towards matching more widespread
    convention of `const T*` rather than `T const*`.
    -   We are aware that some developers have a stylistic preference for "East
        `const`" which would write `T const*` in C++. However, that preference
        and advocacy for this style hasn't yet caused it to become more

>         widespread or widely adopted style.

Reviewer 2: [No action required.] We could even allow both styles, by permitting `const` as either a prefix or postfix operator. This would match C++ better, including weird C++ constructs like `const T* const` (which could also be written as `T const* const` or `const (const T*)` or even `const (T const*)`).

Author: =] That's ... something.

I'm landing this without any action as it's not clear this is worth reviewing more words for this proposal.
